### DISCUSSION TITLE
'Simple Enough' vs. 'Too Simple'

### DISCUSSION THREAD

**Discussion by philj404** [philj404/SimpleSerialShell#20] [General]
Before I created SimpleSerialShell, I looked at several other Arduino shell implementations.  None of them seemed quite what I wanted... so I made this library.

I had these goals in mind... I didn't write them down at the time:

1. Easy to use correctly
    - Working example code
    - Easy to add commands
    - Easy to keep commands near the code they use
    - Minimal setup -- harder to get it wrong
1. Easy to understand
    - Easy to maintain -- existing application code should not need to change
    - Easy to modify
    - Easy to recognize if it won't work for you
1. Minimal dependencies
    - just Arduino and Stream
    - independent of microcontroller used
1. Very small footprint
    - compatible with 2k RAM on Arduino Uno

Later I added some other goals to improve maintainability:
- confirm SimpleSerialShell compiles on different platforms
- Unit tests show the shell executes commands as expected

Basically I still think of these priorities when I consider a code change.  My first release seemed "Simple Enough".

**Comment by brucemack** [comment]
That's a good manifesto. 

I've also looked at several of the choices out there and found this to be the only one that solves the core problem in a Simple Enough way.  

I think the mark of good engineering is that it makes the _**simple things simple, and the complicated things possible**_.  Generally the difference-maker isn't how many features are included (or not), it's how well the header files are structured/commented, how good the README is, how clear the examples are, and how sensible the default behavior is.

**Comment by philj404** [comment]
This change was "Not simple enough"

So... I tried out this change.  Basically I wanted to to move the linked list of Commands into a compile-time table in ROM.  In principle it would take up less RAM space.

The code worked, and _did_ take up less precious RAM... but it was _way_ harder to understand the code:
- Dependencies: the commandTable was in one file, but
  - the commands were either needed to be collected from different files 
  - or commands were disconnected from their local helpers
  - or everything was in one giant file
- Not readable.  Stored strings in Atmel PROGMEM space need special accessors like `strcpy_P()` to differentiate program memory from regular RAM access.  Getting it all stored correctly and then accessed correctly was not intuitive.  It was nowhere near as easy to use as `shell->addCommand(...)`.
- Not backwards compatible.  Supporting legacy applications would mean supporting _both_ the commandTable and `shell->addCommand()`.

I do not want to maintain something so complicated 'forever'.  So I decided to treat those changes as a dead end.  

I learned a lot.  (But now you don't have to learn it too! :grin:)

**Comment by philj404** [comment]
Less Simple Alternatives.

If SimpleSerialShell doesn't do what you need, consider other ways of doing it.  It's OK with me if you find something or invent something better for your needs.

In order from "Simple" to "Definitely not Simple":
- Consider making a code fork of SimpleSerialShell, and modify to fit your needs.  If you don't need to consider support for small memory or other processors, you can be free of a lot of constraints.
- Newer processors are quite powerful and not very expensive.  If you haven't committed to the (excellent and accessible) Arduino ecosystem
  - Consider using MicroPython.  It's based on Python instead of C++, so it's a completely different code base/ecosystem.  But it has its own "REPL" shell built in.
  - Would a RaspberryPi/Linux/bash combination be a good fit?
